packages/nimble-components/src/table/models/keyboard-navigation-manager.ts

```diff
@@ -0,0 +1,1025 @@
+        if (this.hasRowOrCellFocusType() && this.inNavigationMode) {
+            this.focusCurrentRow(false);
+        }
+    }
+
+    private readonly handleFocus = (event: FocusEvent): void => {
+        // User may have clicked elsewhere in the table (on an element not reflected in this.focusState). Update our focusState
+        // based on the current active element in a few cases:
+        // - user is interacting with tabbable content of a cell
+        // - user clicked an action menu. In this case, the active element can either be the MenuButton (no table focus beforehand), or a
+        //   MenuItem (if focus was already in the table when the action menu button was clicked).  If it's a MenuItem, we need to look up
+        //   the linked action menu and cell to figure out what to set our focusState to.
+        // TODO: any better ways to handle this? Only do this if mouse clicked after last key navigation? Cache the active element
+        // as we update focusState, and only do the more complicated logic if activeElement no longer matches what we think it should
+        // be (based on focusState)?
+        const activeElement = this.getActiveElement();
+        let actionMenuActive = false;
+        let row: TableRow | TableGroupRow | undefined;
+        console.log(
+            'table focusin',
+            'target',
+            event.target,
+            'relatedTarget',
+            event.relatedTarget,
+            'active element',
+            this.getActiveElementDebug()
+        );
+        if (activeElement) {
+            row = this.getContainingRow(activeElement);
+            const cell = this.getContainingCell(activeElement);
+            if (row && !(row instanceof TableGroupRow)) {
+                if (cell) {
+                    if (cell.actionMenuButton === activeElement) {
+                        this.focusState.focusType = TableFocusType.cellActionMenu;
+                        this.focusState.rowIndex = row.dataIndex;
+                        this.focusState.columnIndex = this.table.visibleColumns.indexOf(cell.column!);
+                        actionMenuActive = true;
+                        this.inNavigationMode = false;
+                    } else {
+                        const contentIndex = cell.cellView.tabbableChildren.indexOf(
+                            activeElement
+                        );
+                        if (contentIndex > -1) {
+                            this.focusState.focusType = TableFocusType.cellContent;
+                            this.focusState.rowIndex = row.dataIndex;
+                            this.focusState.columnIndex = this.table.visibleColumns.indexOf(cell.column!);
+                            this.focusState.cellContentIndex = contentIndex;
+                            this.inNavigationMode = false;
+                        }
+                    }
+                }
+            }
+        }
+
+        if (event.target instanceof MenuItem) {
```

Comment:
I'm guessing it's not possible to enter into this `if` block if above `if` block was entered? If not, can we make it more clear by making this an `if/else if` block? If it _can_ make sense to enter into both `if` blocks, are there ways for us to avoid duplicate efforts (possibly an over-optimization as iterating over the columns isn't a big deal).